lib/index.js

```diff
@@ -29,34 +29,63 @@ var Klaviyo = module.exports = integration('Klaviyo')
  * @api public
  */
 
-Klaviyo.prototype.identify = function(msg, fn) {
+Klaviyo.prototype.identify = function(mapped, fn) {
   var self = this;
 
-  this
-    .get('/identify')
-    .query({ data: new Buffer(JSON.stringify(msg.peopleData)).toString('base64') })
-    .end(function(err, res){
-      if (err) return fn(err);
-
-      // 1 = success 0 = failure
-      if (res.text != '1') {
-        err = self.error('bad response');
-        return fn(err, res);
-      }
-
-      if (!msg.listData) return fn(null, res);
-
-      // https://www.klaviyo.com/docs/api/lists
-      // Add this person to a List
-      self
-        .post('/v1/list/' + msg.listId + '/members') // upsertion endpoint
-        .type('form')
-        .send(msg.listData)
-        .end(fn);
+  if (!mapped.listId) {
+    return this._identify(mapped.peopleData, fn)
+  }
+
+  var key = [this.settings.apiKey, mapped.peopleData.$id].join(':');
+  this.lock(key, function() {
+    self._addToList(mapped.listId, mapped.listData, function(err, res) {
+      self.unlock(key, function() {
+        if (err) return fn(err);
+        self._identify(mapped.peopleData, fn)
```

Comment:
I believe you should unlock in the callback of this `_identify` function. You shouldn't unlock until all operations are done with